File: KSC2013_ProblemSet_n_Solution/2013년문제답안/Problem1/Q1-serial.c
[문제1번 C 기본 코드]

#include<stdio.h>
#include<stdlib.h>
#include<math.h>
#include<sys/time.h>

int ncall;
typedef enum { FALSE=0, TRUE=1} boolean;
float gettime();

double func(double x){
	ncall ++;
	return 4.L/(x*x+1.L);
}

double simpson0(double (*func1)(double),int n,double aa,double bb){
	double h,xx,rslt;
	boolean lodd;
	int j;

	if((n%2) !=0){
		fprintf(stderr,"input error, n must be even number %d\n",n);
		exit(9);
	}
	rslt = 0.L;
	h = (bb-aa)/(double)n;
	rslt = (func1(aa) + func1(bb));
	for(lodd = TRUE,j = 1; j<n;j++){
		xx = aa + h * (double)j;
		if(lodd) {
			rslt +=  4.L*func1(xx);
		}
		else {
			rslt += 2.L*func1(xx);
		}
		lodd = ((lodd)? (FALSE):(TRUE));
	}
	rslt = rslt*h/3.L;
	return rslt;
}

int main(int argc, char **argv){
	double rslt,aa,bb;
	int n;
	double time_start,time_end;

	time_start = gettime();
	n = 2000000000;
	aa = 0.L;
	bb = 1.L;
	ncall = 0;

	rslt = simpson0(func,n,aa,bb);
	time_end = gettime();
	printf("%d %18.16g n, rslt,%d\n",n,rslt,ncall);
	printf("Wallclock time %g\n",time_end-time_start);
	return 0;
}
struct timeval tv;
float gettime()
{
	static int startflag=1;
	static double tsecs0, tsecs1;	
	if( startflag ) {
		(void ) gettimeofday(&tv,0);
		tsecs0 = tv.tv_sec + tv.tv_usec*1.0e-6;
		startflag = 0;
	}
	(void) gettimeofday(&tv,0);
	tsecs1 = tv.tv_sec + tv.tv_usec*1.0e-6;
	return ((float) (tsecs1-tsecs0));
}

File: KSC2013_ProblemSet_n_Solution/2013년문제답안/Problem1/Q1-serial.f90
[문제1번 포트란 기본 코드]
      module intdata
       implicit none  
       private
       save
       integer ncall
       public :: ncall
       end module intdata

      program simpson
       USE intdata, ONLY : ncall
       implicit none  
       real*8, external :: func 
       real*8 rslt,aa,bb
       integer n
	   integer c1,c2,cr,cm
	   real rate,time_start,time_end
	   call system_clock(count_rate = cr)
	   rate = REAL(cr)
	   call system_clock(c1)
	   time_start = c1/rate
       n=2000000000
       aa=0.0d0
       bb=1.0d0
       ncall=0
       call simpson0(func,n,aa,bb,rslt)
	   call system_clock(c2)
	   time_end = c2/rate
      write(6,*) n,rslt,' n,rslt',ncall
      write(6,*) 'Wallclock time',time_end-time_start
       stop
       end program simpson

       subroutine simpson0(func,n,aa,bb,rslt)
       implicit none
       real*8 func
       integer n
       real*8 rslt,aa,bb
       real*8 h,xx
       integer j
       logical lodd
       rslt=0.d0
       if(mod(n,2) /= 0)then
       print*,' input error, n must be even number',n
       stop
       endif

       h=(bb-aa)/dble(n)
       rslt=(func(aa)+func(bb))
       lodd=.true.
       do j=1,n-1
       xx=aa+h*dble(j)
       if(lodd)then
       rslt=rslt+4.0d0*func(xx)
       else
       rslt=rslt+2.0d0*func(xx)
       endif
       lodd=(.not. lodd)
       enddo

       rslt=rslt*h/3.0d0
       return
       end

       real*8 function func(x)
       USE intdata, ONLY : ncall
       implicit none
       real*8 x
       func=4.d0/(x*x+1.d0)
       ncall=ncall+1
       return
       end
